tests/driver_dfplayer/main.c

```diff
@@ -85,7 +125,7 @@ int main(void)
         "Experiment with the shell command \"dfplayer\"\n\n"
     );
 
-    if (dfplayer_set_callbacks(dfplayer_get(0), _cb_done, _cb_src, NULL)) {
+    if (dfplayer_set_callbacks(dfplayer_get(0), _cb_done, _cb_src, dfplayer_get(0))) {
```

Comment:
The pointer to the `dfplayer_t` struct is already passed to all callback functions already, no?

Reply:
It does indeed make more sense to just pass the device descriptor and safe the RAM of the user pointer. The user would be interested in anything other than the device descriptor anyway.

Reply:
Fixed

Reply:
Having the ability to pass an arbitrary pointer to the callback is still useful - it's just not needed in this example.

Reply:
You can still do this e.g. using:

```C
typedef struct {
    dfplayer_t dfplayer;
    void *userptr;
} dfplayer_with_user_ptr;
```

And now using the `container_of()` macro will give you the user pointer. It has the disadvantage of no longer being able to use `auto_init()`, but it safes 4 byte for most use cases.